src/Analyzers/CSharp/Tests/ForEachCast/ForEachCastTests.cs

```diff
@@ -1043,5 +1046,104 @@ public static void M((object, object)[] x)
             await TestAlwaysAsync(test, code);
             await TestWhenStronglyTypedAsync(test, code);
         }
+
+        [Fact, WorkItem(63470, "https://github.com/dotnet/roslyn/issues/63470")]
+        public async Task TestRegex_GoodCast()
+        {
+            var test = @"
+using System.Text.RegularExpressions;
+
+public static class Program
+{   
+    public static void M(Regex regex, string text)
+    {
+        foreach (Match m in regex.Matches(text))
+        {
+        }
+    }
+}
+";
+            await TestAlwaysAsync(test, test);
+            await TestWhenStronglyTypedAsync(test, test);
+        }
+
+        [Fact, WorkItem(63470, "https://github.com/dotnet/roslyn/issues/63470")]
+        public async Task TestRegex_BadCast()
+        {
+            var test = @"
+using System.Text.RegularExpressions;
+
+public static class Program
+{   
+    public static void M(Regex regex, string text)
+    {
+        [|foreach|] (string m in regex.Matches(text))
+        {
+        }
+    }
+}
+";
+            var code = @"
+using System.Linq;
+using System.Text.RegularExpressions;
+
+public static class Program
+{   
+    public static void M(Regex regex, string text)
+    {
+        foreach (string m in regex.Matches(text).Cast<string>())
+        {
+        }
+    }
+}
+";
+            await TestAlwaysAsync(test, code);
+            await TestWhenStronglyTypedAsync(test, code);
+        }
+
+        [Fact, WorkItem(63470, "https://github.com/dotnet/roslyn/issues/63470")]
+        public async Task WeaklyTypedGetEnumeratorWithIEnumerableOfT()
```

Comment:
It would be good to test the "BadCast" case in this form as well, where the implemented interfaces and GetEnumerator behavior are apparent from the test.